Support TorchScript call over remote API (RRef)

xush6528 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32466 Support TorchScript call over remote API (RRef)**

It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

xush6528 committed:
[WIP] Support TorchScript call over remote API (RRef)

It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` supports taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by make `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 5cc91ea8:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=5cc91ea81e3aa7889b0b80915eac329e2d0d5800).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4310211) pytorch_windows_build (1/1)
**Step:** "Install Cudnn" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=89109252) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=89109252))
<details>
<summary>
<code>cp: target 'C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/' is not a directory</code>
</summary>

```
-- 
Path = cudnn-10.1-windows10-x64-v7.6.4.38.zip 
Type = zip 
Physical Size = 277028683 
 
Everything is Ok 
 
Files: 4 
Size:       422285436 
Compressed: 277028683 
cp: target 'C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v10.1/' is not a directory
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xush6528 committed:
Update on "[WIP] Support TorchScript call over remote API (RRef)"


It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

xush6528 committed:
Update on "Support TorchScript call over remote API (RRef)"


It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

xush6528 committed:
Update on "Support TorchScript call over remote API (RRef)"


It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -50,6 +50,30 @@ c10::intrusive_ptr<c10::ivalue::Future> rpcTorchscriptCall(
+  TORCH_INTERNAL_ASSERT(
+      ctx.getWorkerId() != dst.id_,
+      "Does not support creating RRef on self yet.");
+  auto userRRef = ctx.createUserRRef<IValue>(dst.id_);
```
add type here after #30630 is landed

zhaojuanmao commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -50,6 +50,30 @@ c10::intrusive_ptr<c10::ivalue::Future> rpcTorchscriptCall(
+  ctx.addPendingUser(userRRef->forkId(), userRRef);
+  fm->addCallback(remoteCallCallback);
+
+  return PyRRef(userRRef);
```
we need to return userRRef here and move PyRRef(userRRef) to _invoke_remote_torchscript, as userRRef will be ivalue and then the remoteTorchscript() can be called in JIT.

zhaojuanmao commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -50,6 +50,30 @@ c10::intrusive_ptr<c10::ivalue::Future> rpcTorchscriptCall(
   return futPtr;
 }
 
+PyRRef remoteTorchscript(
```
could we refactor pyRemoteBuiltin and make it to be used by both builtin and torchscript call?

xush6528 replied:
I tried, the inheritance setup about ScripCall, RemoteCall, ScriptCall makes it hard to consolidate in this PR. It's could be another PR's work to streamline them.

zhaojuanmao commented on torch/distributed/rpc/api.py:
```diff
@@ -376,13 +400,16 @@ def remote(to, func, args=None, kwargs=None):
+        return _invoke_remote_builtin(_agent, info, qualified_name, rf, *args, **kwargs)
+    elif isinstance(func, torch.jit.ScriptFunction):
+        return _invoke_remote_torchscript(
+            info, torch._jit_internal._qualified_name(func), *args, **kwargs
```
are you planning to add autograd profiler and test in follow up diff?

xush6528 replied:
Right, I haven't passed the profiler state `rf` in this PR yet. 
I can do it in a following PR. 
I am feeling the 3 code paths are more compact this way.

zhaojuanmao commented on torch/distributed/rpc/api.py:
```diff
@@ -357,6 +363,24 @@ def remote(to, func, args=None, kwargs=None):
+        >>>    return torch.add(t1, t2)
+        >>> import torch.distributed.rpc as rpc
+        >>> rpc.init_rpc("worker0", rank=0, world_size=2)
+        >>> rref = rpc.remote("worker1", my_script_add, args=(torch.ones(2), 3))
```
we also need _remote() version that accepts qualified name only, so that it can be consistent with JIT interface later on. more specifically, the JIT interface will be private temporarily because JIT ops do not accept callable yet, for now we plan to have a private _remote() to accept qualified_name

zhaojuanmao reviewed (commented):
overall looks good! thanks for working on this. also, would you please add a test in rpc_test.py that can show the remote script call can receive remote exception?

mrshenli commented on torch/distributed/rpc/api.py:
```diff
@@ -324,7 +329,8 @@ def remote(to, func, args=None, kwargs=None):
     Arguments:
         to (str or WorkerInfo): id or name of the destination worker.
-        func (callable): builtin functions (like :meth:`torch.add`).
+        func (callable): any callable function. python callable, builtin or annotated TorchScript
```
good catch, thanks!

mrshenli commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -50,6 +50,30 @@ c10::intrusive_ptr<c10::ivalue::Future> rpcTorchscriptCall(
   return futPtr;
 }
 
+PyRRef remoteTorchscript(
```
Any reason one function is called `rpcTorchscript*Call*` and another is called `remoteTorchscript`?

xush6528 replied:
I'll remove the word, "Call", to make it in align with function names in "python_functions.h".

mrshenli commented on torch/csrc/distributed/rpc/script_functions.cpp:
```diff
@@ -50,6 +50,30 @@ c10::intrusive_ptr<c10::ivalue::Future> rpcTorchscriptCall(
+      *agent, dst, std::move(*scriptRemoteCall).toMessage(), false, nullptr);
+
+  ctx.addPendingUser(userRRef->forkId(), userRRef);
+  fm->addCallback(remoteCallCallback);
```
I guess this is only reason for moving `remoteCallCallback` out of an anonymous namespace? Since this is from a different file, let's give it a more informative name. What about `confirmPendingUser`/`confirmUser`?

xush6528 replied:
That's a better name.

mrshenli commented on torch/csrc/distributed/rpc/python_functions.h:
```diff
@@ -41,6 +41,10 @@ PyRRef pyRemotePythonUdf(
     std::vector<torch::Tensor>& tensors,
     const std::shared_ptr<torch::autograd::profiler::RecordFunction>& rf);
 
+void remoteCallCallback(
```
I am not sure if this function fit in here. All other functions are exposed to Python as RPC entry points to C++, but not this one. Maybe expose it on RRefContext? Or maybe we can tolerate a little code duplication here?

xush6528 replied:
Moved to rref_context.h/cpp, and put it into a "callback" namespace.

xush6528 committed:
Update on "Support TorchScript call over remote API (RRef)"


It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

xush6528 committed:
Update on "Support TorchScript call over remote API (RRef)"


It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

xush6528 commented:
Update to resolve comments. Will wait for and rebase after #30630.

xush6528 committed:
Update on "Support TorchScript call over remote API (RRef)"


It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

xush6528 committed:
Update on "Support TorchScript call over remote API (RRef)"


It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/rpc/request_callback_impl.cpp:
```diff
@@ -80,16 +79,26 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
 
-      TypePtr ret_type = src.op()->schema().returns()[0].type();
-      auto ownerRRef = ctx.getOrCreateOwnerRRef(src.retRRefId(), ret_type);
+      TypePtr ret_type = scriptRemoteCall.op()->schema().returns()[0].type();
```
nit: returnType

zhaojuanmao commented on torch/distributed/rpc/api.py:
```diff
@@ -680,3 +706,60 @@ def _rpc_async_torchscript(to, qualified_name, args=None, kwargs=None):
+
+    Arguments:
+        to (str or WorkerInfo): id or name of the destination worker.
+        func (callable): any callable function. python callable, builtin or annotated TorchScript
```
nit: string, not callable

zhaojuanmao commented on torch/distributed/rpc/api.py:
```diff
@@ -680,3 +706,60 @@ def _rpc_async_torchscript(to, qualified_name, args=None, kwargs=None):
+@_require_initialized
+def _remote_torchscript(to, qualified_name, args=None, kwargs=None):
+    r"""
+    Make a remote call to run ``func`` on worker ``to`` and return an
```
nit run TorchScript func

zhaojuanmao reviewed (approved):
lgtm, please address inline comments and make sure all tests are green before landing

zhaojuanmao commented:
test failures are real

xush6528 committed:
Update on "Support TorchScript call over remote API (RRef)"


It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

xush6528 committed:
Update on "Support TorchScript call over remote API (RRef)"


It's a follow-up work of https://github.com/pytorch/pytorch/pull/32197.

In https://github.com/pytorch/pytorch/pull/32197, `rpc.sync_rpc(..) `and `rpc.rpc_async(..)` support taking a TorchScript annotated Python function as the user function for RPC.

This PR extend along this direction by making `rpc.remote(..)` support taking a TorchScript annotated Python function as well.

Differential Revision: [D19440633](https://our.internmc.facebook.com/intern/diff/D19440633/)

[ghstack-poisoned]

xush6528 commented:
The 2 failures are irrelevant to distributed module.
